diff -crB hid-xpadneo/dkms.post_install hid-xpadneo.new/dkms.post_install
*** hid-xpadneo/dkms.post_install	2023-03-01 02:13:11.483605532 -0300
--- hid-xpadneo.new/dkms.post_install	2023-03-01 03:21:07.876809434 -0300
***************
*** 21,33 ****
  		echo "Y" >"${DISABLE_ERTM}"
  	fi
  fi
- 
- echo "Installing modalias database..."
- install -D -m 0644 -t /etc/modprobe.d etc-modprobe.d/xpadneo.conf
- 
- echo "Installing udev rules..."
- install -D -m 0644 -t /etc/udev/rules.d etc-udev-rules.d/50-xpadneo-fixup-steamlink.rules
- install -D -m 0644 -t /etc/udev/rules.d etc-udev-rules.d/60-xpadneo.rules
- 
- echo "Reloading udev..."
- udevadm control --reload
--- 21,23 ----
diff -crB hid-xpadneo/dkms.post_remove hid-xpadneo.new/dkms.post_remove
*** hid-xpadneo/dkms.post_remove	2023-03-01 02:13:11.483605532 -0300
--- hid-xpadneo.new/dkms.post_remove	2023-03-01 03:21:11.027791943 -0300
***************
*** 4,16 ****
  ERTM_OVERRIDE="/etc/modprobe.d/99-xpadneo-bluetooth.conf"
  
  [ -L "${ERTM_OVERRIDE}" ] || { echo "Uninstalling ERTM override..."; rm -f "${ERTM_OVERRIDE}"; }
- 
- echo "Uninstalling modalias database..."
- rm -f /etc/modprobe.d/xpadneo.conf
- 
- echo "Uninstalling udev rules..."
- rm -f /etc/udev/rules.d/60-xpadneo.rules
- rm -f /etc/udev/rules.d/50-xpadneo-fixup-steamlink.rules
- 
- echo "Reloading udev..."
- udevadm control --reload
--- 4,6 ----
