diff -aur hid-xpadneo/dkms.conf.in hid-xpadneo.new/dkms.conf.in
--- hid-xpadneo/dkms.conf.in	2022-11-29 23:43:54.472158200 -0300
+++ hid-xpadneo.new/dkms.conf.in	2023-03-01 03:29:28.753210314 -0300
@@ -1,5 +1,5 @@
 PACKAGE_NAME="hid-xpadneo"
-PACKAGE_VERSION="@DO_NOT_CHANGE@"
+PACKAGE_VERSION="@PACKAGE_VERSION@"
 BUILT_MODULE_NAME[0]="hid-xpadneo"
 BUILT_MODULE_LOCATION[0]="src"
 DEST_MODULE_LOCATION[0]="/kernel/drivers/hid"
