From 1053a6988291f79d7ce6143f57e7add1098fc6e0 Mon Sep 17 00:00:00 2001
From: Bouke Haarsma <boukehaarsma23@gmail.com>
Date: Sun, 23 Jul 2023 21:54:26 +0200
Subject: [PATCH] Use k10-temp power sensors if exist

---
 src/cpu.cpp | 22 ++++++++++++++++++++++
 src/cpu.h   |  6 ++++++
 2 files changed, 28 insertions(+)

diff --git a/src/cpu.cpp b/src/cpu.cpp
index a338739..1619c65 100644
--- a/src/cpu.cpp
+++ b/src/cpu.cpp
@@ -296,6 +296,21 @@ bool CPUStats::UpdateCpuTemp() {
 static bool get_cpu_power_k10temp(CPUPowerData* cpuPowerData, float& power) {
     CPUPowerData_k10temp* powerData_k10temp = (CPUPowerData_k10temp*)cpuPowerData;
 
+    if(powerData_k10temp->corePowerFile || powerData_k10temp->socPowerFile)
+    {
+        rewind(powerData_k10temp->corePowerFile);
+        rewind(powerData_k10temp->socPowerFile);
+        fflush(powerData_k10temp->corePowerFile);
+        fflush(powerData_k10temp->socPowerFile);
+        int corePower, socPower;
+        if (fscanf(powerData_k10temp->corePowerFile, "%d", &corePower) != 1)
+            goto voltagebased;
+        if (fscanf(powerData_k10temp->socPowerFile, "%d", &socPower) != 1)
+            goto voltagebased;
+        power = (corePower + socPower) / 1000000;
+        return true;
+    }
+    voltagebased:
     if (!powerData_k10temp->coreVoltageFile || !powerData_k10temp->coreCurrentFile || !powerData_k10temp->socVoltageFile || !powerData_k10temp->socCurrentFile)
         return false;
 
@@ -494,21 +509,28 @@ static CPUPowerData_k10temp* init_cpu_power_data_k10temp(const std::string path)
 
     std::string coreVoltageInput, coreCurrentInput;
     std::string socVoltageInput, socCurrentInput;
+    std::string socPowerInput, corePowerInput;
 
     if(!find_input(path, "in", coreVoltageInput, "Vcore")) return nullptr;
     if(!find_input(path, "curr", coreCurrentInput, "Icore")) return nullptr;
     if(!find_input(path, "in", socVoltageInput, "Vsoc")) return nullptr;
     if(!find_input(path, "curr", socCurrentInput, "Isoc")) return nullptr;
+    if(!find_input(path, "power", corePowerInput, "Pcore")) return nullptr;
+    if(!find_input(path, "power", socPowerInput, "Psoc")) return nullptr;
 
     SPDLOG_DEBUG("hwmon: using input: {}", coreVoltageInput);
     SPDLOG_DEBUG("hwmon: using input: {}", coreCurrentInput);
     SPDLOG_DEBUG("hwmon: using input: {}", socVoltageInput);
     SPDLOG_DEBUG("hwmon: using input: {}", socCurrentInput);
+    SPDLOG_DEBUG("hwmon: using input: {}", corePowerInput);
+    SPDLOG_DEBUG("hwmon: using input: {}", socPowerInput);
 
     powerData->coreVoltageFile = fopen(coreVoltageInput.c_str(), "r");
     powerData->coreCurrentFile = fopen(coreCurrentInput.c_str(), "r");
     powerData->socVoltageFile = fopen(socVoltageInput.c_str(), "r");
     powerData->socCurrentFile = fopen(socCurrentInput.c_str(), "r");
+    powerData->socPowerFile = fopen(socPowerInput.c_str(), "r");
+    powerData->corePowerFile = fopen(corePowerInput.c_str(), "r");
 
     return powerData.release();
 }
diff --git a/src/cpu.h b/src/cpu.h
index ed91d98..ef31735 100644
--- a/src/cpu.h
+++ b/src/cpu.h
@@ -72,12 +72,18 @@ struct CPUPowerData_k10temp : public CPUPowerData {
          fclose(this->socVoltageFile);
       if(this->socCurrentFile)
          fclose(this->socCurrentFile);
+      if(this->corePowerFile)
+         fclose(this->corePowerFile);
+      if(this->socPowerFile)
+         fclose(this->socPowerFile);
    };
 
    FILE* coreVoltageFile {nullptr};
    FILE* coreCurrentFile {nullptr};
    FILE* socVoltageFile {nullptr};
    FILE* socCurrentFile {nullptr};
+   FILE* corePowerFile {nullptr};
+   FILE* socPowerFile {nullptr};
 };
 
 struct CPUPowerData_zenpower : public CPUPowerData {
-- 
2.41.0

