diff --git a/src/gallium/frontends/rusticl/meson.build b/src/gallium/frontends/rusticl/meson.build
--- a/src/gallium/frontends/rusticl/meson.build
+++ b/src/gallium/frontends/rusticl/meson.build
@@ -92,7 +92,6 @@
 rusticl_bindgen_args = [
   '--no-convert-floats',
   '--use-array-pointers-in-arguments',
-  '--size_t-is-usize',
   '--default-enum-style', 'rust',
   '--with-derive-partialeq',
   '--with-derive-eq',
